File: src/dart/lib/src/renewal.dart
import 'lease.dart';

/// Logical-clock domain shared with JavaScript clients.
const int maxRenewalClockMs = 9007199254740991;

/// Largest TTL exactly representable by every supported duration type.
const int maxRenewalTtlMs = 9223372036854;

/// Scheduling policy for a held Fiducia lease.
final class RenewalPolicy {
  final Duration renewEvery;
  final Duration safetyMargin;

  const RenewalPolicy({required this.renewEvery, required this.safetyMargin});

  static const defaults = RenewalPolicy(
    renewEvery: Duration(seconds: 20),
    safetyMargin: Duration(seconds: 10),
  );
}

/// Why this supervisor permanently stopped authorizing protected effects.
enum RenewalLossReason {
  invalidPolicy('invalid_policy'),
  clockRegression('clock_regression'),
  expired('expired'),
  renewalFailed('renewal_failed'),
  identityChanged('identity_changed'),
  tokenChanged('token_changed'),
  deadlineMissing('deadline_missing'),
  deadlineInvalid('deadline_invalid'),
  deadlineRegressed('deadline_regressed'),
  completionAfterDeadline('completion_after_deadline'),
  deadlineOverflow('deadline_overflow'),
  invalidTtl('invalid_ttl');

  final String wire;
  const RenewalLossReason(this.wire);
}

/// Terminal renewal error. Once recorded by a supervisor it is sticky.
final class RenewalError implements Exception {
  final RenewalLossReason reason;
  final String message;
  final Object? cause;

  const RenewalError(this.reason, this.message, {this.cause});

  @override
  String toString() => cause == null
      ? '${reason.wire}: $message'
      : '${reason.wire}: $message; cause: $cause';
}

sealed class RenewalDecision {
  const RenewalDecision();
}

final class RenewalWait extends RenewalDecision {
  final int checkInMs;
  const RenewalWait(this.checkInMs);
}

final class RenewalNow extends RenewalDecision {
  const RenewalNow();
}

final class RenewalLost extends RenewalDecision {
  final RenewalLossReason reason;
  const RenewalLost(this.reason);
}

sealed class RenewalCheckpoint {
  const RenewalCheckpoint();
}

final class RenewalCheckpointWait extends RenewalCheckpoint {
  final int checkInMs;
  const RenewalCheckpointWait(this.checkInMs);
}

final class RenewalCheckpointRenewed extends RenewalCheckpoint {
  final int checkInMs;
  const RenewalCheckpointRenewed(this.checkInMs);
}

/// Source of process-local monotonic milliseconds.
typedef MonotonicClock = int Function();

/// A [Stopwatch]-based monotonic clock suitable for one supervisor lifetime.
final class StopwatchClock {
  final Stopwatch _stopwatch;
  final int _originMs;

  StopwatchClock({int originMs = 0})
    : _originMs = originMs,
      _stopwatch = Stopwatch()..start();

  int call() => _originMs + _stopwatch.elapsedMilliseconds;
}

/// Sticky, cooperative supervisor for one exact Fiducia grant identity.
///
/// Call [checkpoint] before every authoritative commit. This cannot undo an
/// unfenced side effect already emitted by the caller; protected datastores
/// must still admit the fencing token atomically with the mutation.
final class LeaseRenewalSupervisor {
  LeaseGrant _grant;
  final RenewalPolicy policy;
  int _localDeadlineMs;
  int _nextRenewalMs;
  int _lastObservedMs;
  RenewalError? _loss;

  LeaseRenewalSupervisor(LeaseGrant grant, this.policy, int nowMs)
    : _grant = grant,
      _localDeadlineMs = 0,
      _nextRenewalMs = 0,
      _lastObservedMs = nowMs {
    _validateGrant(grant);
    _validateClock(nowMs);
    _validateAuthorityDeadline(grant.leaseExpiresMs);
    final schedule = _schedule(nowMs, grant.ttlMs, policy);
    _localDeadlineMs = schedule.$1;
    _nextRenewalMs = schedule.$2;
  }

  LeaseGrant get grant => _grant;
  int get localDeadlineMs => _localDeadlineMs;
  int get nextRenewalMs => _nextRenewalMs;
  RenewalError? get loss => _loss;
  bool get isLive => _loss == null;

  RenewalDecision decide(int nowMs) {
    final loss = _loss;
    if (loss != null) return RenewalLost(loss.reason);
    try {
      _validateClock(nowMs);
    } on RenewalError catch (error) {
      return _lose(error);
    }
    if (nowMs < _lastObservedMs) {
      return _lose(
        RenewalError(
          RenewalLossReason.clockRegression,
          'monotonic clock regressed from $_lastObservedMs to $nowMs',
        ),
      );
    }
    _lastObservedMs = nowMs;
    if (nowMs >= _localDeadlineMs) {
      return _lose(
        RenewalError(
          RenewalLossReason.expired,
          'local lease deadline $_localDeadlineMs was reached at $nowMs',
        ),
      );
    }
    if (nowMs >= _nextRenewalMs) return const RenewalNow();
    return RenewalWait(_nextRenewalMs - nowMs);
  }

  void assertLive(int nowMs) {
    if (decide(nowMs) case RenewalLost()) throw _loss!;
  }

  Future<RenewalCheckpoint> checkpoint(
    Lease lease, {
    required MonotonicClock clock,
  }) async {
    final decision = decide(clock());
    if (decision is RenewalWait) {
      return RenewalCheckpointWait(decision.checkInMs);
    }
    if (decision is RenewalLost) throw _loss!;

    final previous = _grant;
    late final LeaseGrant renewed;
    try {
      renewed = await lease.renew(
        previous,
        Duration(milliseconds: previous.ttlMs),
      );
    } catch (cause) {
      throw _fail(
        RenewalError(
          RenewalLossReason.renewalFailed,
          'lease authority did not prove continued ownership',
          cause: cause,
        ),
      );
    }
    return acceptRenewal(clock(), renewed);
  }

  RenewalCheckpoint acceptRenewal(int completedMs, LeaseGrant renewed) {
    if (_loss != null) throw _loss!;
    try {
      _validateClock(completedMs);
      _validateGrant(renewed);
    } on RenewalError catch (error) {
      throw _fail(error);
    }
    if (completedMs < _lastObservedMs) {
      throw _fail(
        RenewalError(
          RenewalLossReason.clockRegression,
          'monotonic clock regressed from $_lastObservedMs to $completedMs during renewal',
        ),
      );
    }
    _lastObservedMs = completedMs;
    if (completedMs >= _localDeadlineMs) {
      throw _fail(
        RenewalError(
          RenewalLossReason.completionAfterDeadline,
          'renewal completed at $completedMs, not before local deadline $_localDeadlineMs',
        ),
      );
    }
    if (renewed.key != _grant.key || renewed.holder != _grant.holder) {
      throw _fail(
        const RenewalError(
          RenewalLossReason.identityChanged,
          'renewal changed the lock key or holder identity',
        ),
      );
    }
    if (renewed.fencingToken != _grant.fencingToken) {
      throw _fail(
        RenewalError(
          RenewalLossReason.tokenChanged,
          'renewal changed fencing token ${_grant.fencingToken} to ${renewed.fencingToken}',
        ),
      );
    }
    try {
      _validateDeadlineProgress(_grant.leaseExpiresMs, renewed.leaseExpiresMs);
      final schedule = _schedule(completedMs, renewed.ttlMs, policy);
      _grant = renewed;
      _localDeadlineMs = schedule.$1;
      _nextRenewalMs = schedule.$2;
      return RenewalCheckpointRenewed(_nextRenewalMs - completedMs);
    } on RenewalError catch (error) {
      throw _fail(error);
    }
  }

  RenewalDecision _lose(RenewalError error) {
    final recorded = _record(error);
    return RenewalLost(recorded.reason);
  }

  RenewalError _fail(RenewalError error) => _record(error);

  RenewalError _record(RenewalError error) {
    _loss ??= error;
    return _loss!;
  }
}

void _validateGrant(LeaseGrant grant) {
  if (grant.holder.isEmpty) {
    throw const RenewalError(
      RenewalLossReason.identityChanged,
      'grant holder must be non-empty',
    );
  }
  if (grant.fencingToken.isNegative ||
      grant.fencingToken > BigInt.parse('18446744073709551615')) {
    throw const RenewalError(
      RenewalLossReason.tokenChanged,
      'fencing token must be an unsigned 64-bit integer',
    );
  }
  if (grant.ttlMs <= 0 || grant.ttlMs > maxRenewalTtlMs) {
    throw const RenewalError(
      RenewalLossReason.invalidTtl,
      'lease TTL is outside the shared millisecond domain',
    );
  }
  _validateAuthorityDeadline(grant.leaseExpiresMs);
}

void _validateClock(int value) {
  if (value < 0 || value > maxRenewalClockMs) {
    throw const RenewalError(
      RenewalLossReason.deadlineOverflow,
      'logical clock is outside the shared millisecond domain',
    );
  }
}

void _validateAuthorityDeadline(int? value) {
  if (value != null && (value <= 0 || value > maxRenewalClockMs)) {
    throw const RenewalError(
      RenewalLossReason.deadlineInvalid,
      'authority deadline must be positive and within the shared millisecond domain',
    );
  }
}

void _validateDeadlineProgress(int? previous, int? renewed) {
  _validateAuthorityDeadline(renewed);
  if (previous != null && renewed == null) {
    throw const RenewalError(
      RenewalLossReason.deadlineMissing,
      'renewal omitted a deadline that the authority previously reported',
    );
  }
  if (previous != null && renewed != null && renewed <= previous) {
    throw RenewalError(
      RenewalLossReason.deadlineRegressed,
      'authority deadline did not advance: $previous -> $renewed',
    );
  }
}

(int, int) _schedule(int nowMs, int ttlMs, RenewalPolicy policy) {
  _validateClock(nowMs);
  if (ttlMs <= 0 || ttlMs > maxRenewalTtlMs) {
    throw const RenewalError(
      RenewalLossReason.invalidTtl,
      'lease TTL is outside the shared millisecond domain',
    );
  }
  final renewEveryMs = policy.renewEvery.inMilliseconds;
  final safetyMarginMs = policy.safetyMargin.inMilliseconds;
  if (renewEveryMs <= 0 ||
      safetyMarginMs <= 0 ||
      renewEveryMs >= ttlMs ||
      safetyMarginMs >= ttlMs) {
    throw const RenewalError(
      RenewalLossReason.invalidPolicy,
      'renewal interval and safety margin must both be positive and less than the lease TTL',
    );
  }
  final deadlineMs = nowMs + ttlMs;
  final intervalDue = nowMs + renewEveryMs;
  if (deadlineMs > maxRenewalClockMs || intervalDue > maxRenewalClockMs) {
    throw const RenewalError(
      RenewalLossReason.deadlineOverflow,
      'renewal schedule exceeds the shared logical-clock domain',
    );
  }
  final marginDue = deadlineMs - safetyMarginMs;
  final nextRenewalMs = intervalDue < marginDue ? intervalDue : marginDue;
  if (nextRenewalMs <= nowMs) {
    throw const RenewalError(
      RenewalLossReason.invalidPolicy,
      'renewal policy leaves no positive live interval',
    );
  }
  return (deadlineMs, nextRenewalMs);
}

File: src/rust/renewal.rs
//! Deterministic, fail-closed supervision for long-running Fiducia leases.
//!
//! A renewal supervisor is cooperative: callers checkpoint it before every
//! authoritative commit. It does not make an unfenced external side effect
//! safe and it cannot undo effects that were emitted before a checkpoint.
//! Protected datastores must still admit the grant's fencing token atomically
//! with the mutation.

use std::error::Error;
use std::fmt;
use std::time::Duration;

use crate::{Lease, LeaseGrant, LockError};

/// Maximum logical millisecond value shared safely with browser runtimes.
pub const MAX_RENEWAL_CLOCK_MS: u64 = 9_007_199_254_740_991;

/// Maximum TTL representable without narrowing in every supported runtime.
/// Go and Dart both ultimately use signed 64-bit duration storage.
pub const MAX_RENEWAL_TTL_MS: u64 = 9_223_372_036_854;

/// Scheduling policy for one held grant.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct RenewalPolicy {
    /// Preferred interval between successful renewal checkpoints.
    pub renew_every: Duration,
    /// Latest acceptable window before the local deadline.
    pub safety_margin: Duration,
}

impl RenewalPolicy {
    pub const fn new(renew_every: Duration, safety_margin: Duration) -> Self {
        Self {
            renew_every,
            safety_margin,
        }
    }

    pub fn renew_every_ms(self) -> u64 {
        duration_ms(self.renew_every)
    }

    pub fn safety_margin_ms(self) -> u64 {
        duration_ms(self.safety_margin)
    }
}

impl Default for RenewalPolicy {
    fn default() -> Self {
        Self::new(Duration::from_secs(20), Duration::from_secs(10))
    }
}

/// Why this supervisor permanently stopped authorizing protected effects.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum RenewalLossReason {
    InvalidPolicy,
    ClockRegression,
    Expired,
    RenewalFailed,
    IdentityChanged,
    TokenChanged,
    DeadlineMissing,
    DeadlineInvalid,
    DeadlineRegressed,
    CompletionAfterDeadline,
    DeadlineOverflow,
    InvalidTtl,
}

impl RenewalLossReason {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::InvalidPolicy => "invalid_policy",
            Self::ClockRegression => "clock_regression",
            Self::Expired => "expired",
            Self::RenewalFailed => "renewal_failed",
            Self::IdentityChanged => "identity_changed",
            Self::TokenChanged => "token_changed",
            Self::DeadlineMissing => "deadline_missing",
            Self::DeadlineInvalid => "deadline_invalid",
            Self::DeadlineRegressed => "deadline_regressed",
            Self::CompletionAfterDeadline => "completion_after_deadline",
            Self::DeadlineOverflow => "deadline_overflow",
            Self::InvalidTtl => "invalid_ttl",
        }
    }
}

impl fmt::Display for RenewalLossReason {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(self.as_str())
    }
}

/// Terminal renewal error. Once stored by a supervisor, it is sticky.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RenewalError {
    pub reason: RenewalLossReason,
    pub message: String,
    pub cause: Option<LockError>,
}

impl RenewalError {
    fn new(reason: RenewalLossReason, message: impl Into<String>) -> Self {
        Self {
            reason,
            message: message.into(),
            cause: None,
        }
    }

    fn caused_by(reason: RenewalLossReason, message: impl Into<String>, cause: LockError) -> Self {
        Self {
            reason,
            message: message.into(),
            cause: Some(cause),
        }
    }
}

impl fmt::Display for RenewalError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match &self.cause {
            Some(cause) => write!(
                formatter,
                "{}: {}; cause: {cause}",
                self.reason, self.message
            ),
            None => write!(formatter, "{}: {}", self.reason, self.message),
        }
    }
}

impl Error for RenewalError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        self.cause
            .as_ref()
            .map(|cause| cause as &(dyn Error + 'static))
    }
}

/// Pure scheduling decision at one monotonic-clock observation.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RenewalDecision {
    Wait { check_in_ms: u64 },
    RenewNow,
    Lost { reason: RenewalLossReason },
}

/// Result of a successful asynchronous checkpoint.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RenewalCheckpoint {
    Wait { check_in_ms: u64 },
    Renewed { check_in_ms: u64 },
}

/// Source of process-local monotonic milliseconds.
pub trait MonotonicClock {
    fn now_ms(&self) -> u64;
}

impl<F> MonotonicClock for F
where
    F: Fn() -> u64,
{
    fn now_ms(&self) -> u64 {
        self()
    }
}

/// Sticky supervisor for one exact Fiducia grant identity.
#[derive(Debug, Clone)]
pub struct RenewalSupervisor {
    grant: LeaseGrant,
    policy: RenewalPolicy,
    local_deadline_ms: u64,
    next_renewal_ms: u64,
    last_observed_ms: u64,
    loss: Option<RenewalError>,
}

impl RenewalSupervisor {
    /// Start supervision at `now_ms`, measured by a monotonic process clock.
    pub fn new(
        grant: LeaseGrant,
        policy: RenewalPolicy,
        now_ms: u64,
    ) -> Result<Self, RenewalError> {
        validate_clock(now_ms)?;
        validate_authority_deadline(grant.lease_expires_ms)?;
        let (local_deadline_ms, next_renewal_ms) = schedule(now_ms, grant.ttl_ms, policy)?;
        Ok(Self {
            grant,
            policy,
            local_deadline_ms,
            next_renewal_ms,
            last_observed_ms: now_ms,
            loss: None,
        })
    }

    pub fn grant(&self) -> &LeaseGrant {
        &self.grant
    }

    pub const fn local_deadline_ms(&self) -> u64 {
        self.local_deadline_ms
    }

    pub const fn next_renewal_ms(&self) -> u64 {
        self.next_renewal_ms
    }

    pub fn loss(&self) -> Option<&RenewalError> {
        self.loss.as_ref()
    }

    pub fn is_live(&self) -> bool {
        self.loss.is_none()
    }

    /// Observe the clock and decide whether to wait, renew, or stop.
    pub fn decision(&mut self, now_ms: u64) -> RenewalDecision {
        if let Some(loss) = &self.loss {
            return RenewalDecision::Lost {
                reason: loss.reason,
            };
        }
        if let Err(error) = validate_clock(now_ms) {
            return self.lose(error);
        }
        if now_ms < self.last_observed_ms {
            return self.lose(RenewalError::new(
                RenewalLossReason::ClockRegression,
                format!(
                    "monotonic clock regressed from {} to {now_ms}",
                    self.last_observed_ms
                ),
            ));
        }
        self.last_observed_ms = now_ms;
        if now_ms >= self.local_deadline_ms {
            return self.lose(RenewalError::new(
                RenewalLossReason::Expired,
                format!(
                    "local lease deadline {} was reached at {now_ms}",
                    self.local_deadline_ms
                ),
            ));
        }
        if now_ms >= self.next_renewal_ms {
            RenewalDecision::RenewNow
        } else {
            RenewalDecision::Wait {
                check_in_ms: self.next_renewal_ms - now_ms,
            }
        }
    }

    /// Fail unless authority is still live at this checkpoint.
    pub fn assert_live(&mut self, now_ms: u64) -> Result<(), RenewalError> {
        match self.decision(now_ms) {
            RenewalDecision::Lost { .. } => Err(self
                .loss
                .clone()
                .expect("lost decision always records a terminal error")),
            RenewalDecision::Wait { .. } | RenewalDecision::RenewNow => Ok(()),
        }
    }

    /// Invoke `Lease::renew` only when due. Any failed or ambiguous renewal is
    /// terminal; subsequent checkpoints return the original sticky loss and do
    /// not contact the authority again.
    pub async fn checkpoint<L, C>(
        &mut self,
        lease: &L,
        clock: &C,
    ) -> Result<RenewalCheckpoint, RenewalError>
    where
        L: Lease + Sync,
        C: MonotonicClock,
    {
        match self.decision(clock.now_ms()) {
            RenewalDecision::Wait { check_in_ms } => {
                return Ok(RenewalCheckpoint::Wait { check_in_ms });
            }
            RenewalDecision::Lost { .. } => {
                return Err(self
                    .loss
                    .clone()
                    .expect("lost decision always records a terminal error"));
            }
            RenewalDecision::RenewNow => {}
        }

        let previous = self.grant.clone();
        let renewed = match lease
            .renew(&previous, Duration::from_millis(previous.ttl_ms))
            .await
        {
            Ok(renewed) => renewed,
            Err(cause) => {
                let error = RenewalError::caused_by(
                    RenewalLossReason::RenewalFailed,
                    "lease authority did not prove continued ownership",
                    cause,
                );
                self.lose(error);
                return Err(self.loss.clone().expect("loss was just recorded"));
            }
        };

        let completed_ms = clock.now_ms();
        self.accept_renewal(completed_ms, renewed)
    }

    /// Validate one successful authority response and advance the local window.
    pub fn accept_renewal(
        &mut self,
        completed_ms: u64,
        renewed: LeaseGrant,
    ) -> Result<RenewalCheckpoint, RenewalError> {
        if let Some(loss) = &self.loss {
            return Err(loss.clone());
        }
        if let Err(error) = validate_clock(completed_ms) {
            return Err(self.record(error));
        }
        if completed_ms < self.last_observed_ms {
            return Err(self.record(RenewalError::new(
                RenewalLossReason::ClockRegression,
                format!(
                    "monotonic clock regressed from {} to {completed_ms} during renewal",
                    self.last_observed_ms
                ),
            )));
        }
        self.last_observed_ms = completed_ms;
        if completed_ms >= self.local_deadline_ms {
            return Err(self.record(RenewalError::new(
                RenewalLossReason::CompletionAfterDeadline,
                format!(
                    "renewal completed at {completed_ms}, not before local deadline {}",
                    self.local_deadline_ms
                ),
            )));
        }
        if renewed.key != self.grant.key || renewed.holder != self.grant.holder {
            return Err(self.record(RenewalError::new(
                RenewalLossReason::IdentityChanged,
                "renewal changed the lock key or holder identity",
            )));
        }
        if renewed.fencing_token != self.grant.fencing_token {
            return Err(self.record(RenewalError::new(
                RenewalLossReason::TokenChanged,
                format!(
                    "renewal changed fencing token {} to {}",
                    self.grant.fencing_token, renewed.fencing_token
                ),
            )));
        }
        if let Err(error) =
            validate_deadline_progress(self.grant.lease_expires_ms, renewed.lease_expires_ms)
        {
            return Err(self.record(error));
        }
        let (local_deadline_ms, next_renewal_ms) =
            match schedule(completed_ms, renewed.ttl_ms, self.policy) {
                Ok(schedule) => schedule,
                Err(error) => return Err(self.record(error)),
            };
        self.grant = renewed;
        self.local_deadline_ms = local_deadline_ms;
        self.next_renewal_ms = next_renewal_ms;
        Ok(RenewalCheckpoint::Renewed {
            check_in_ms: next_renewal_ms - completed_ms,
        })
    }

    fn lose(&mut self, error: RenewalError) -> RenewalDecision {
        let reason = self.record(error).reason;
        RenewalDecision::Lost { reason }
    }

    fn record(&mut self, error: RenewalError) -> RenewalError {
        if self.loss.is_none() {
            self.loss = Some(error);
        }
        self.loss.clone().expect("terminal loss is present")
    }
}

fn duration_ms(duration: Duration) -> u64 {
    u64::try_from(duration.as_millis()).unwrap_or(u64::MAX)
}

fn validate_clock(now_ms: u64) -> Result<(), RenewalError> {
    if now_ms <= MAX_RENEWAL_CLOCK_MS {
        Ok(())
    } else {
        Err(RenewalError::new(
            RenewalLossReason::DeadlineOverflow,
            format!("logical clock {now_ms} exceeds {MAX_RENEWAL_CLOCK_MS}"),
        ))
    }
}

fn validate_authority_deadline(deadline: Option<u64>) -> Result<(), RenewalError> {
    match deadline {
        Some(0) => Err(RenewalError::new(
            RenewalLossReason::DeadlineInvalid,
            "authority deadline must be positive when present",
        )),
        Some(value) if value > MAX_RENEWAL_CLOCK_MS => Err(RenewalError::new(
            RenewalLossReason::DeadlineOverflow,
            format!("authority deadline {value} exceeds {MAX_RENEWAL_CLOCK_MS}"),
        )),
        _ => Ok(()),
    }
}

fn validate_deadline_progress(
    previous: Option<u64>,
    renewed: Option<u64>,
) -> Result<(), RenewalError> {
    validate_authority_deadline(renewed)?;
    match (previous, renewed) {
        (Some(_), None) => Err(RenewalError::new(
            RenewalLossReason::DeadlineMissing,
            "renewal omitted a deadline that the authority previously reported",
        )),
        (Some(previous), Some(next)) if next <= previous => Err(RenewalError::new(
            RenewalLossReason::DeadlineRegressed,
            format!("authority deadline did not advance: {previous} -> {next}"),
        )),
        _ => Ok(()),
    }
}

fn schedule(now_ms: u64, ttl_ms: u64, policy: RenewalPolicy) -> Result<(u64, u64), RenewalError> {
    if ttl_ms == 0 || ttl_ms > MAX_RENEWAL_TTL_MS {
        return Err(RenewalError::new(
            RenewalLossReason::InvalidTtl,
            format!("lease TTL must be within 1..={MAX_RENEWAL_TTL_MS} ms"),
        ));
    }
    let renew_every_ms = policy.renew_every_ms();
    let safety_margin_ms = policy.safety_margin_ms();
    if renew_every_ms == 0
        || safety_margin_ms == 0
        || renew_every_ms >= ttl_ms
        || safety_margin_ms >= ttl_ms
    {
        return Err(RenewalError::new(
            RenewalLossReason::InvalidPolicy,
            "renewal interval and safety margin must both be positive and less than the lease TTL",
        ));
    }
    let deadline_ms = now_ms
        .checked_add(ttl_ms)
        .filter(|value| *value <= MAX_RENEWAL_CLOCK_MS)
        .ok_or_else(|| {
            RenewalError::new(
                RenewalLossReason::DeadlineOverflow,
                "local lease deadline exceeds the shared logical-clock domain",
            )
        })?;
    let interval_due = now_ms.checked_add(renew_every_ms).ok_or_else(|| {
        RenewalError::new(
            RenewalLossReason::DeadlineOverflow,
            "renewal interval overflowed the logical clock",
        )
    })?;
    let margin_due = deadline_ms - safety_margin_ms;
    let next_renewal_ms = interval_due.min(margin_due);
    if next_renewal_ms <= now_ms {
        return Err(RenewalError::new(
            RenewalLossReason::InvalidPolicy,
            "renewal policy leaves no positive live interval",
        ));
    }
    Ok((deadline_ms, next_renewal_ms))
}
